### ISSUE TITLE
How to get the code better documented?

### ISSUE THREAD

**Issue by JohannesLorenz** [LMMS/lmms#4932]
@Spekular requested more documentation (or documentation *at all*) in our code. I can agree. But how do we assure that?

Our idea was to require coders to document those classes they work with. E.g. when I worked on the Lv2 plugin, I had to learn how the LMMS `Plugin` class works. Though I only changed a few code lines in it, I ended up [documenting the whole Plugin class](https://github.com/LMMS/lmms/commit/a1b355828e347b8e7fca8118873fc834953c1d67#diff-1840771aa9cc1d07324ea9a8059c8291). How to ensure that? We could simply add this to the [code review conventions](https://github.com/LMMS/lmms/wiki/Reviewing-Pull-Requests).

Any opinions, or other suggestions?

**Comment by Spekular** [comment]
I think comment conventions could be added as a subsection at the bottom of the coding conventions wiki page, perhaps? Could include examples of good/bad comments, whether we do `// This` or `//This`, which methods need/don't need comments, and so on.

**Comment by musikBear** [comment]
convention ed 
// are for in-method short remarks
/**/ are meant to be larger in depth explaining content *in-between* blocks of code
I would like a 'header' comment on each and every method, ( -and ofcause class), it should be clear where the method get input from, and where the processed output is used, kind of a 'red-tread' through the code.
Comments should preferably be understandable, even for rookies, so they need serious insight! 
( The rhein-gold would be a *proper design documentation*, but that train has (sadly) departed

**Comment by darkmastermindz** [comment]
> @Spekular requested more documentation (or documentation _at all_) in our code. I can agree. But how do we assure that?
> 
> Our idea was to require coders to document those classes they work with. E.g. when I worked on the Lv2 plugin, I had to learn how the LMMS `Plugin` class works. Though I only changed a few code lines in it, I ended up [documenting the whole Plugin class](https://github.com/LMMS/lmms/commit/a1b355828e347b8e7fca8118873fc834953c1d67#diff-1840771aa9cc1d07324ea9a8059c8291). How to ensure that? We could simply add this to the [code review conventions](https://github.com/LMMS/lmms/wiki/Reviewing-Pull-Requests).
> 
> Any opinions, or other suggestions?

I think all code functions should contain an explanation above with appropriate tags as a standard to help autogenerate documentation easier see: https://developer.lsst.io/cpp/api-docs.html

**Comment by he29-net** [comment]
@darkmastermindz Does anyone actually read the autogenerated documentation? I never do.

For me, it just adds noise that makes the actual comment less readable. Not to mention it is somewhat annoying to write (i.e. what's the syntax this week? Was it Doxygen? Or was it something else?).

And while I definitely do support documenting inputs and outputs of functions, not every single one of them must have it (getters/setters are obvious, as may be some simple functions with well named parameters), so forcing it may simply add _too many_ comments where they may not be needed.

**Comment by darkmastermindz** [comment]
> @darkmastermindz Does anyone actually read the autogenerated documentation? I never do.
> 
> For me, it just adds noise that makes the actual comment less readable. Not to mention it is somewhat annoying to write (i.e. what's the syntax this week? Was it Doxygen? Or was it something else?).
> 
> And while I definitely do support documenting inputs and outputs of functions, not every single one of them must have it (getters/setters are obvious, as may be some simple functions with well named parameters), so forcing it may simply add _too many_ comments where they may not be needed.

I agree there isn't a need to document everything because you can have self explanatory code like getters and setters. But it's useful to have more complex functions because if someone uses and IDE, it'll show the documentation while using the function. Of course, this probably can be better achieved using and ADTs / Doxygen to set some standards that can show up in something like IntelliSense. Within open source, you don't have the luxury of assuming that other developers looking at your code are familiar with the context in which it was written or its history. So, I'd approach commenting with short/consiceness "Can I guess what this function does within 5 minutes if I'm new to this project?"

**Comment by JohannesLorenz** [comment]
> But it's useful to have more complex functions because if someone uses and IDE, it'll show the documentation while using the function.

IMO that's a very good point for using doxygen.
